### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by mrmonday** [libpnet/libpnet#2]
Currently all the code is written for Windows support, but it does not work:
 * On Win32 receives time out
 * On Win64 the tests segfault

These issues need investigating and fixing.

**Comment by cybergeek94** [comment]
Win64 is no longer segfaulting. It's just hanging like Win32.

**Comment by mrmonday** [comment]
See also pull #41. Thanks!

**Comment by cybergeek94** [comment]
Something just struck me as odd. All the examples I find of using the Packet API show allocating the packet, calling the read function, and freeing the packet all in the same thread--the same function, actually. But our tests create the packet in one thread and send it to another to be read into; perhaps the hanging is coming from dependency on thread-local data or some other thread insafety in the API? I couldn't find any documentation on it but all uses I see definitely show the setup and teardown in the same thread as the reading.

**Comment by mrmonday** [comment]
Looking at the definition of `PacketAllocatePacket` - http://dog.tele.jp/winpcap/html/Packet32_8c-source.html#l00686 - it seems it simply allocates some memory and returns it. It uses an unusual function which I haven't seen before, so following the trail...

The initial call:
```c
GlobalAllocPtr(GMEM_MOVEABLE | GMEM_ZEROINIT,sizeof(PACKET));
```
Definition of `GlobalAllocPtr` from `windowsx.h`:
```c
#define GlobalAllocPtr(flags,cb) (GlobalLock(GlobalAlloc((flags),(cb))))
```
Documentation for `GlobalLock` and `GlobalAlloc`:
 * `GlobalLock`: https://msdn.microsoft.com/en-us/library/windows/desktop/aa366584%28v=vs.85%29.aspx
 * `GlobalAlloc`: https://msdn.microsoft.com/en-us/library/windows/desktop/aa366574%28v=vs.85%29.aspx

Looking through this information leads me to think that it should be safe to allocate the memory in one thread, then access it in another. It might be worth investigating though - you could try skipping the libpnet API and writing out a quick test file which just calls those functions directly, or maybe using `malloc

**Comment by cybergeek94** [comment]
Guess what?! [I was right!][1]

```
$ cargo test
   Compiling pnet v0.0.1 (file:///C:/Users/Austin/Rust/libpnet)
     Running target\pnet-60a2dc8adb9158df.exe

running 12 tests
test old_packet::ethernet::ethernet_header_test ... ok
test old_packet::ipv4::ipv4_header_test ... ok
test old_packet::ipv6::ipv6_header_test ... ok
test old_packet::udp::udp_header_ipv4_test ... ok
test old_packet::udp::udp_header_ipv6_test ... ok
test test::check_test_environment ... ok
thread '<unnamed>' panicked at 'layer2 failed: a device attached to the system is not functioning.
', src\test.rs:317
test test::layer2 ... FAILED
test test::layer3_ipv4 ... FAILED
test test::layer4_ipv4 ... FAILED
test test::layer4_ipv6 ... FAILED
test test::packet_api_test ... ok
test util::mac_addr_from_str ... ok

failures:

---- test::layer2 stdout ----
        thread 'test::layer2' panicked at 'layer2 test thread panicked! layer2 failed: a device attached to the system is not functioning.
', src\test.rs:332


---- test::layer3_ipv4 stdout ----
        thread 'test::layer3_ipv4' panicked at 'layer3: unable to create channel: either the application has not called wsastartup, or wsastartup failed.
', src\test.rs:204


---- test::layer4_ipv4 stdout ----
        thread 'test::layer4_ipv4' panicked at 'layer4: unable to create channel: either the application has not called wsastartup, or wsastartup failed.
', src\test.rs:140


---- test::layer4_ipv6 stdout ----
        thread 'test::layer4_ipv6' panicked at 'layer4: unable to create channel: either the application has not called wsastartup, or wsastartup failed.
', src\test.rs:140



failures:
    test::layer2
    test::layer3_ipv4
    test::layer4_ipv4
    test::layer4_ipv6

test result: FAILED. 8 passed; 4 failed; 0 ignored; 0 measured

thread '<main>' panicked at 'Some tests failed', C:\bot\slave\nightly-dist-rustc-win-64\build\src\libtest\lib.rs:256
```

Thus, the Packet API is *not* thread-safe. I'm thinking that the adapter has to be opened in the same thread that it's read from. I don't think that's a big problem, though. Just don't open the adapter until it's first read or written to, and have a different adapter instance for each end of the channel for `datalink_channel`.

[1]: https://github.com/cybergeek94/libpnet/commit/b3dde7751204e3d169e3bdd388f955d083f3eba9

**Comment by mrmonday** [comment]
Great work, good find! I'm very excited to see the test suite passing on Windows, seems like you're getting close :)

**Comment by mrmonday** [comment]
I've added support for AppVeyor now, so there's CI for Win32. It would seem I broke the Windows build with some of my recent changes, I won't fix it though since I figure it'll make things a pain for you to merge.

I'd quite like to get it testing on Win64 as well, but one thing at a time!

**Comment by cybergeek94** [comment]
There's something I want to experiment with. The `layer_2` test writes to and from the same adapter handle in two separate threads; that can't be good. Perhaps opening a separate handle for the reader and writer sides would be sufficient?

**Comment by mrmonday** [comment]
Definitely worth a shot, yeah.

**Comment by cybergeek94** [comment]
Damn, just tried it. No dice. Separate handle for reading and writing, and I even tried commenting out the call to `DataLinkSenderImpl::send_to()` to see if it was just blocking the device. I just don't think it likes sending adapter handles across threads; it will have to be initialized in a thread-local.

**Comment by cybergeek94** [comment]
Actually, thread-local isn't necessary. We should open the adapter in `DataLinkReceiverImpl::iter()`. As for the sending side... I'm not sure. I haven't gotten a chance to experience possible thread-insafety with that side of it.

**Comment by mrmonday** [comment]
I guess the thing to do is:

1. Check that this will actually fix this - make the send/receive tuple have `Option` fields for the adapter, then throw in a quick `if None = self.adapter` and initialise it in the relevant send/receive functions. This should be enough to have the adapter created in the correct thread without too much effort.
2. If it works, see if there's a way to do static initializers for threads created with std::thread/task/whatever it is these days - that way we can avoid the overhead of checking for each send/receive operation.

Another possible angle to tackle this from would be to take a look to see how WinPcap handles this, since I assume using the libpcap abstraction does not cause this problem.

**Comment by cybergeek94** [comment]
I fixed it, in this case. You know what it was? `read_buffer_size` was too small. The example I found set it to 512K but the test set it to only 128.

**Comment by mrmonday** [comment]
There needs to be a real solution to the `*_buffer_size` problem, we have had sooo many issues caused due to it being the incorrect size, or we've used the wrong one or something...

**Comment by cybergeek94** [comment]
From the docs for `PacketSetBuff`:
> The buffer size is set to 0 when an instance of the driver is opened: the programmer should remember to set it to a proper value. As an example, wpcap sets the buffer size to 1MB at the beginning of a capture. 

The logic error looks to be a problem with a 0-length packet.

**Comment by cybergeek94** [comment]
Update: the `layer_2` test now passes with the sender logic uncommented, even sharing the same adapter handle across threads. `DataLinkChannelIteratorImpl::next()` needs to handle 0-length packets as it appears that's what happens on a timeout instead of an IoError.

**Comment by mrmonday** [comment]
Awesome, great work!

I think we should remove the call to set a timeout - none of the other backends timeout, I merely added it as a debugging step (thought it seems it has caused greater confusion...)

It might be useful to support timeouts later, but for now let's not bother.

**Comment by mrmonday** [comment]
Windows support now works, as of 0.4.0. See pull #121.
